Remora.Discord.Commands/API/ResponseTrackingInteractionAPI.cs

```diff
@@ -0,0 +1,212 @@
+//
+//  ResponseTrackingInteractionAPI.cs
+//
+//  Author:
+//       Jarl Gullberg <jarl.gullberg@gmail.com>
+//
+//  Copyright (c) 2017 Jarl Gullberg
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU Lesser General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using OneOf;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Abstractions.Rest;
+using Remora.Discord.Commands.Contexts;
+using Remora.Discord.Commands.Services;
+using Remora.Rest.Core;
+using Remora.Results;
+
+namespace Remora.Discord.Commands.API;
+
+/// <summary>
+/// Implements a tracking system for whether an interaction has been responded to.
+/// </summary>
+internal class ResponseTrackingInteractionAPI : IDiscordRestInteractionAPI
+{
+    private readonly ContextInjectionService? _contextInjector;
+    private readonly IDiscordRestInteractionAPI _actual;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ResponseTrackingInteractionAPI"/> class.
+    /// </summary>
+    /// <param name="actual">The actual API instance.</param>
+    /// <param name="provider">The service provider.</param>
+    public ResponseTrackingInteractionAPI
+    (
+        IDiscordRestInteractionAPI actual,
+        IServiceProvider provider
+    )
+    {
+        _actual = actual;
+
+        // TODO: Improve this with a proper scope check once I figure out how to do that
```

Comment:
Should it not just work by having it as constructor parameter?
As the scoped ContextInjectionService will be nevertheless will be injected each time in the transient IDiscordRestInteractionAPI.
Or do you actually want to detect if the REST API is being instantiated inside of a scope?

Reply:
Unfortunately, using a straight constructor parameter doesn't work when resolving the API instance from the root service provider, which is still a use case I want to support. If there's no context in play, the service should still be usable as-is.